Performance improvements by re-using record buffer

harshavardhana opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
Performance improvements by re-using record buffer
<!--- Describe your changes in detail -->

## Motivation and Context
Avoid unnecessary pointer reference allocations
when not needed, for example

- *SelectFuncs{}
- *Row{}

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## Regression
No
<!-- Is this PR fixing a regression? (Yes / No) -->
<!-- If Yes, optionally please include minio version or commit id or PR# that caused this regression, if you have these details. -->

## How Has This Been Tested?
- Test this PR with a large enough CSV file like around 100G, avoids memory allocations are done inline in the CSV reader.
- `mc select --query "select * from S3Object" myminio/testbucket/large.csv`
- We have to closely observe the memory usage. 
- Other parts in the PR is theoretical fix basically avoiding pointer references to improve GC performance over long period of time. https://medium.com/@vCabbage/go-are-pointers-a-performance-optimization-a95840d3ef85 this is due to a problem how the copying pointers causing heap usage. 

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

donatello commented on pkg/s3select/input.go:
```diff
@@ -151,7 +151,7 @@ func NewInput(opts *Options) (*Input, error) {
 	reader.firstRow = nil
 
-	reader.reader.FieldsPerRecord = -1
+	reader.reader.FieldsPerRecord = 0
```
This would be a bug when csv files have different numbers of fields in each line.

harshavardhana replied:
I think S3 is not supporting that AFAIK

donatello replied:
No, it is supported - you can try with FileHeaderInfo = None and access columns by index (`s._3` etc). If the input file has rows with different numbers of fields, this will work with S3 but not with this PR.

harshavardhana replied:
Then I think we can make sure that if Header is set we don't need to honor variadic records.

harshavardhana committed:
Performance improvements by re-using record buffer

Avoid unnecessary pointer reference allocations
when not needed, for example

- *SelectFuncs{}
- *Row{}

sinhaashish reviewed (dismissed):
LGTM , tested with some queries

donatello commented on pkg/s3select/input.go:
```diff
@@ -106,22 +119,20 @@ func New(gr io.Reader, size int64, req ObjectSelectRequest) (s3s format.Select,
 func Execute(writer io.Writer, f format.Select) error {
-	myRow := make(chan *Row)
-	curBuf := bytes.NewBuffer(make([]byte, 1000000))
+	myRow := make(chan Row, 10000)
```
This looks like a large accumulation in memory of results (1MiB/row * 10K rows = 10GiB) in case the rows are consumed slowly.

harshavardhana replied:
Oh yes you are right. Let me change it back to 1000 instead.

donatello reviewed (changes requested):
LGTM except for the size of a bounded channel in the code. 

Performance improvement appears to be marginal in my testing (about 0.8% faster than master for csv files with 1MiB sized rows on simple select * like query) - not much visible different in memory usage to me.

sinhaashish reviewed (approved):
LGTM
